File: experiments/ab-trading/core/modules/module_registry.py
#!/usr/bin/env python3
"""
WorkBuddy OS 模块注册表加载器 (Python 侧)

位置: experiments/ab-trading/core/modules/module_registry.py

功能:
1. 从 YAML 文件加载模块注册表（双端唯一真相源）
2. JSON Schema 校验
3. 内存缓存 + 索引加速
4. 热更新（文件监听）
5. 统一查询接口（按ID/链/分类/标签/阶段等）

设计原则:
- 单一真相源：1-ARCHITECTURE/registry/module_registry.yaml
- 只读加载：注册表是配置，运行时不修改
- 索引加速：按多种维度建立索引，查询O(1)
- 降级容错：加载失败时使用内置空注册表，不影响核心流程
"""

import os
import json
import time
import threading
from pathlib import Path
from typing import Dict, List, Optional, Any, Set
from dataclasses import dataclass, field

import yaml

try:
    import jsonschema
    _HAS_JSONSCHEMA = True
except ImportError:
    _HAS_JSONSCHEMA = False


_PROJECT_ROOT = Path(__file__).parent.parent.parent.parent.parent
REGISTRY_PATH = _PROJECT_ROOT / "1-ARCHITECTURE" / "registry" / "module_registry.yaml"
SCHEMA_PATH = _PROJECT_ROOT / "1-ARCHITECTURE" / "registry" / "module_registry.schema.json"


@dataclass
class ModuleInfo:
    """模块信息（扁平化结构，便于查询）"""
    id: str
    name: str
    description: str
    version: str
    chain: str
    category: str
    tags: List[str]
    lifecycle: Dict
    security_level: str
    estimated_tokens: int
    estimated_latency_ms: int
    confidence_range: List[float]
    applicable_stages: List[str]
    applicable_intents: List[str]
    market_conditions: List[str]
    historical_accuracy: float
    historical_calls: int
    dependencies: List[str]
    adapter: Dict
    fallback: Dict
    domain: str = ""
    category_name: str = ""

    @classmethod
    def from_dict(cls, data: Dict, domain: str = "", category_name: str = "") -> "ModuleInfo":
        return cls(
            id=data["id"],
            name=data["name"],
            description=data["description"],
            version=data["version"],
            chain=data["chain"],
            category=data["category"],
            tags=data.get("tags", []),
            lifecycle=data.get("lifecycle", {}),
            security_level=data.get("security_level", "R1"),
            estimated_tokens=data.get("estimated_tokens", 0),
            estimated_latency_ms=data.get("estimated_latency_ms", 0),
            confidence_range=data.get("confidence_range", [0, 100]),
            applicable_stages=data.get("applicable_stages", []),
            applicable_intents=data.get("applicable_intents", []),
            market_conditions=data.get("market_conditions", []),
            historical_accuracy=data.get("historical_accuracy", 0),
            historical_calls=data.get("historical_calls", 0),
            dependencies=data.get("dependencies", []),
            adapter=data.get("adapter", {}),
            fallback=data.get("fallback", {}),
            domain=domain,
            category_name=category_name,
        )

    def to_dict(self) -> Dict:
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "version": self.version,
            "chain": self.chain,
            "category": self.category,
            "tags": self.tags,
            "lifecycle": self.lifecycle,
            "security_level": self.security_level,
            "estimated_tokens": self.estimated_tokens,
            "estimated_latency_ms": self.estimated_latency_ms,
            "confidence_range": self.confidence_range,
            "applicable_stages": self.applicable_stages,
            "applicable_intents": self.applicable_intents,
            "market_conditions": self.market_conditions,
            "historical_accuracy": self.historical_accuracy,
            "historical_calls": self.historical_calls,
            "dependencies": self.dependencies,
            "adapter": self.adapter,
            "fallback": self.fallback,
            "domain": self.domain,
            "category_name": self.category_name,
        }


class ModuleRegistry:
    """
    模块注册表加载器
    
    负责加载、校验、索引和查询模块注册表
    线程安全，支持热更新
    """

    def __init__(self, registry_path: Optional[Path] = None,
                 schema_path: Optional[Path] = None,
                 auto_watch: bool = False):
        self._registry_path = registry_path or REGISTRY_PATH
        self._schema_path = schema_path or SCHEMA_PATH
        
        self._lock = threading.RLock()
        self._modules: Dict[str, ModuleInfo] = {}
        self._raw_data: Dict = {}
        self._last_load_time: float = 0
        self._last_file_mtime: float = 0
        
        self._by_chain: Dict[str, Set[str]] = {}
        self._by_category: Dict[str, Set[str]] = {}
        self._by_domain: Dict[str, Set[str]] = {}
        self._by_stage: Dict[str, Set[str]] = {}
        self._by_tag: Dict[str, Set[str]] = {}
        
        self._watch_thread: Optional[threading.Thread] = None
        self._watch_running: bool = False
        self._watch_interval: float = 5.0
        
        self._load()
        
        if auto_watch:
            self.start_watching()

    def _load(self) -> bool:
        """加载注册表文件并构建索引"""
        try:
            if not self._registry_path.exists():
                print(f"[ModuleRegistry] 注册表文件不存在: {self._registry_path}")
                return False
            
            file_mtime = self._registry_path.stat().st_mtime
            if file_mtime == self._last_file_mtime and self._modules:
                return True
            
            with open(self._registry_path, 'r', encoding='utf-8') as f:
                raw_data = yaml.safe_load(f)
            
            if not raw_data or not isinstance(raw_data, dict):
                print("[ModuleRegistry] 注册表格式错误")
                return False
            
            if _HAS_JSONSCHEMA and self._schema_path.exists():
                try:
                    with open(self._schema_path, 'r', encoding='utf-8') as f:
                        schema = json.load(f)
                    jsonschema.validate(instance=raw_data, schema=schema)
                except jsonschema.ValidationError as e:
                    print(f"[ModuleRegistry] Schema校验失败: {e}")
                except Exception as e:
                    print(f"[ModuleRegistry] Schema校验异常: {e}")
            
            modules: Dict[str, ModuleInfo] = {}
            domains = raw_data.get("domains", {})
            
            for domain_key, domain_data in domains.items():
                categories = domain_data.get("categories", {})
                for cat_key, cat_data in categories.items():
                    mod_map = cat_data.get("modules", {})
                    for mod_key, mod_data in mod_map.items():
                        info = ModuleInfo.from_dict(
                            mod_data,
                            domain=domain_key,
                            category_name=cat_key
                        )
                        modules[info.id] = info
            
            with self._lock:
                self._raw_data = raw_data
                self._modules = modules
                self._last_load_time = time.time()
                self._last_file_mtime = file_mtime
                self._build_indexes()
            
            print(f"[ModuleRegistry] 加载成功，共 {len(modules)} 个模块")
            return True
            
        except Exception as e:
            print(f"[ModuleRegistry] 加载失败: {e}")
            import traceback
            traceback.print_exc()
            return False

    def _build_indexes(self) -> None:
        """构建查询索引（调用方需持有锁）"""
        self._by_chain.clear()
        self._by_category.clear()
        self._by_domain.clear()
        self._by_stage.clear()
        self._by_tag.clear()
        
        for mid, mod in self._modules.items():
            if mod.chain not in self._by_chain:
                self._by_chain[mod.chain] = set()
            self._by_chain[mod.chain].add(mid)
            
            if mod.category not in self._by_category:
                self._by_category[mod.category] = set()
            self._by_category[mod.category].add(mid)
            
            if mod.domain not in self._by_domain:
                self._by_domain[mod.domain] = set()
            self._by_domain[mod.domain].add(mid)
            
            for stage in mod.applicable_stages:
                if stage not in self._by_stage:
                    self._by_stage[stage] = set()
                self._by_stage[stage].add(mid)
            
            for tag in mod.tags:
                if tag not in self._by_tag:
                    self._by_tag[tag] = set()
                self._by_tag[tag].add(mid)

    def reload(self) -> bool:
        """强制重新加载"""
        return self._load()

    def check_for_updates(self) -> bool:
        """检查文件是否有更新，有则重新加载"""
        try:
            if not self._registry_path.exists():
                return False
            file_mtime = self._registry_path.stat().st_mtime
            if file_mtime > self._last_file_mtime:
                print(f"[ModuleRegistry] 检测到文件更新，重新加载...")
                return self._load()
            return False
        except Exception:
            return False

    def start_watching(self, interval: float = 5.0) -> None:
        """启动文件监听线程"""
        if self._watch_running:
            return
        
        self._watch_interval = interval
        self._watch_running = True
        self._watch_thread = threading.Thread(
            target=self._watch_loop,
            daemon=True,
            name="ModuleRegistryWatcher"
        )
        self._watch_thread.start()
        print(f"[ModuleRegistry] 文件监听已启动，间隔 {interval}s")

    def stop_watching(self) -> None:
        """停止文件监听"""
        self._watch_running = False
        if self._watch_thread:
            self._watch_thread.join(timeout=2.0)
            self._watch_thread = None

    def _watch_loop(self) -> None:
        """监听循环"""
        while self._watch_running:
            try:
                self.check_for_updates()
            except Exception:
                pass
            time.sleep(self._watch_interval)

    def get(self, module_id: str) -> Optional[ModuleInfo]:
        """根据ID获取模块"""
        with self._lock:
            return self._modules.get(module_id)

    def has(self, module_id: str) -> bool:
        """检查模块是否存在"""
        with self._lock:
            return module_id in self._modules

    def get_all(self) -> List[ModuleInfo]:
        """获取所有模块"""
        with self._lock:
            return list(self._modules.values())

    def count(self) -> int:
        """获取模块总数"""
        with self._lock:
            return len(self._modules)

    def query(self,
              chain: Optional[str] = None,
              category: Optional[str] = None,
              domain: Optional[str] = None,
              stage: Optional[str] = None,
              tag: Optional[str] = None,
              security_level: Optional[str] = None,
              min_accuracy: Optional[float] = None,
              max_tokens: Optional[int] = None,
              intent: Optional[str] = None,
              market_condition: Optional[str] = None) -> List[ModuleInfo]:
        """
        多条件查询模块
        
        Args:
            chain: 所属链 (A/C/F/G/T)
            category: 分类
            domain: 领域
            stage: 适用阶段
            tag: 标签
            security_level: 安全等级
            min_accuracy: 最低历史准确率
            max_tokens: 最大token消耗
            intent: 适用意图
            market_condition: 适用市场条件
            
        Returns:
            匹配的模块列表
        """
        with self._lock:
            candidates = set(self._modules.keys())
            
            if chain and chain in self._by_chain:
                candidates &= self._by_chain[chain]
            
            if category and category in self._by_category:
                candidates &= self._by_category[category]
            
            if domain and domain in self._by_domain:
                candidates &= self._by_domain[domain]
            
            if stage and stage in self._by_stage:
                candidates &= self._by_stage[stage]
            
            if tag and tag in self._by_tag:
                candidates &= self._by_tag[tag]
            
            results = []
            for mid in candidates:
                mod = self._modules[mid]
                
                if security_level and mod.security_level != security_level:
                    continue
                
                if min_accuracy is not None and mod.historical_accuracy < min_accuracy:
                    continue
                
                if max_tokens is not None and mod.estimated_tokens > max_tokens:
                    continue
                
                if intent and intent not in mod.applicable_intents:
                    continue
                
                if market_condition and market_condition not in mod.market_conditions:
                    continue
                
                results.append(mod)
            
            return results

    def get_by_chain(self, chain: str) -> List[ModuleInfo]:
        """按链查询"""
        return self.query(chain=chain)

    def get_by_domain(self, domain: str) -> List[ModuleInfo]:
        """按领域查询"""
        return self.query(domain=domain)

    def get_dependencies(self, module_id: str) -> List[ModuleInfo]:
        """获取模块的依赖"""
        mod = self.get(module_id)
        if not mod:
            return []
        result = []
        for dep_id in mod.dependencies:
            dep = self.get(dep_id)
            if dep:
                result.append(dep)
        return result

    def get_fallback(self, module_id: str) -> Optional[ModuleInfo]:
        """获取模块的降级模块"""
        mod = self.get(module_id)
        if not mod or not mod.fallback.get("enabled"):
            return None
        fallback_id = mod.fallback.get("fallback_module")
        if not fallback_id:
            return None
        return self.get(fallback_id)

    def get_raw(self) -> Dict:
        """获取原始YAML数据"""
        with self._lock:
            return self._raw_data

    def get_domains(self) -> List[str]:
        """获取所有领域"""
        with self._lock:
            return list(self._by_domain.keys())

    def get_chains(self) -> List[str]:
        """获取所有链"""
        with self._lock:
            return list(self._by_chain.keys())

    def get_categories(self) -> List[str]:
        """获取所有分类"""
        with self._lock:
            return list(self._by_category.keys())

    def get_stats(self) -> Dict:
        """获取统计信息"""
        with self._lock:
            stats = {
                "total": len(self._modules),
                "by_chain": {k: len(v) for k, v in self._by_chain.items()},
                "by_domain": {k: len(v) for k, v in self._by_domain.items()},
                "by_category": {k: len(v) for k, v in self._by_category.items()},
                "last_load": self._last_load_time,
            }
            return stats

    def is_active(self, module_id: str) -> bool:
        """检查模块是否处于活跃状态"""
        mod = self.get(module_id)
        if not mod:
            return False
        return mod.lifecycle.get("status") == "active" and not mod.lifecycle.get("deprecated", False)


_global_registry: Optional[ModuleRegistry] = None
_global_lock = threading.Lock()


def get_module_registry(auto_watch: bool = False) -> ModuleRegistry:
    """获取全局模块注册表单例"""
    global _global_registry
    if _global_registry is None:
        with _global_lock:
            if _global_registry is None:
                _global_registry = ModuleRegistry(auto_watch=auto_watch)
    return _global_registry


def reload_registry() -> bool:
    """重新加载全局注册表"""
    registry = get_module_registry()
    return registry.reload()


__all__ = [
    "ModuleInfo",
    "ModuleRegistry",
    "get_module_registry",
    "reload_registry",
    "REGISTRY_PATH",
    "SCHEMA_PATH",
]

File: 3.1-FRONTEND/src/lib/planner/module-registry.ts
/**
 * ============================================================
 *  WorkBuddy OS 模块注册表加载器 (TS 侧)
 * ============================================================
 *
 * 位置: 6-图结构上下文压缩/planner/module-registry.ts
 *
 * 功能:
 * 1. 从 JSON 文件加载模块注册表（双端唯一真相源）
 * 2. 内存缓存 + 索引加速
 * 3. 热更新（文件监听，Node.js 环境）
 * 4. 统一查询接口（按ID/链/分类/标签/阶段等）
 *
 * 设计原则:
 * - 单一真相源: 1-ARCHITECTURE/registry/module_registry.json
 * - 只读加载: 注册表是配置，运行时不修改
 * - 索引加速: 按多种维度建立索引，查询O(1)
 * - 执行器分离: 只加载元数据，执行逻辑保持在现有 SkillsRegistry
 */

import * as fs from 'fs';
import * as path from 'path';

// ============================================================
// 类型定义
// ============================================================

export interface Lifecycle {
  status: 'active' | 'inactive' | 'deprecated' | 'experimental';
  phase: 'prototype' | 'beta' | 'production' | 'eol';
  deprecated: boolean;
  replaced_by: string | null;
}

export interface Adapter {
  type: 'skill' | 'api' | 'local' | 'external' | 'mcp';
  execution_engine: 'python' | 'typescript' | 'both' | 'external';
  skill_md?: string;
  base_url?: string;
  endpoint?: string;
  ts_path?: string;
  py_path?: string;
  py_module?: string;
  skill_path?: string;
  factory_function?: string;
}

export interface Fallback {
  enabled: boolean;
  fallback_module?: string;
  fallback_type?: string;
  fallback_reason?: string;
}

export interface ModuleInfo {
  id: string;
  name: string;
  description: string;
  version: string;
  chain: 'A' | 'C' | 'F' | 'G' | 'T';
  category: string;
  tags: string[];
  lifecycle: Lifecycle;
  security_level: 'R0' | 'R1' | 'R2' | 'R3';
  estimated_tokens: number;
  estimated_latency_ms: number;
  confidence_range: [number, number];
  applicable_stages: Array<'research' | 'analysis' | 'design' | 'validate' | 'execute'>;
  applicable_intents: string[];
  market_conditions: string[];
  historical_accuracy: number;
  historical_calls: number;
  dependencies: string[];
  adapter: Adapter;
  fallback: Fallback;
  domain: string;
  category_name: string;
}

export interface ModuleRegistryRaw {
  version: string;
  updated_at: string;
  schema_version: string;
  total_modules: number;
  domains: Record<string, {
    name: string;
    description: string;
    color: string;
    categories: Record<string, {
      name: string;
      description: string;
      modules: Record<string, any>;
    }>;
  }>;
}

export interface ModuleQueryParams {
  chain?: ModuleInfo['chain'] | ModuleInfo['chain'][];
  category?: string | string[];
  domain?: string | string[];
  stage?: ModuleInfo['applicable_stages'][number] | ModuleInfo['applicable_stages'][number][];
  tag?: string | string[];
  security_level?: ModuleInfo['security_level'];
  min_accuracy?: number;
  max_tokens?: number;
  intent?: string;
  market_condition?: string;
}

// ============================================================
// 注册表路径
// ============================================================

const PROJECT_ROOT = path.resolve(__dirname, '..', '..');
const REGISTRY_JSON_PATH = path.join(
  PROJECT_ROOT,
  '1-ARCHITECTURE',
  'registry',
  'module_registry.json'
);

// ============================================================
// 模块注册表加载器
// ============================================================

export class ModuleRegistry {
  private modules: Map<string, ModuleInfo> = new Map();
  private rawData: ModuleRegistryRaw | null = null;
  private lastLoadTime: number = 0;
  private lastFileMtime: number = 0;

  private byChain: Map<string, Set<string>> = new Map();
  private byCategory: Map<string, Set<string>> = new Map();
  private byDomain: Map<string, Set<string>> = new Map();
  private byStage: Map<string, Set<string>> = new Map();
  private byTag: Map<string, Set<string>> = new Map();

  private watchInterval: ReturnType<typeof setInterval> | null = null;
  private registryPath: string;

  constructor(registryPath?: string) {
    this.registryPath = registryPath || REGISTRY_JSON_PATH;
    this.load();
  }

  // ============================================================
  // 加载与刷新
  // ============================================================

  load(): boolean {
    try {
      if (!fs.existsSync(this.registryPath)) {
        console.warn(`[ModuleRegistry] 注册表文件不存在: ${this.registryPath}`);
        return false;
      }

      const stat = fs.statSync(this.registryPath);
      const fileMtime = stat.mtimeMs;
      if (fileMtime === this.lastFileMtime && this.modules.size > 0) {
        return true;
      }

      const raw = fs.readFileSync(this.registryPath, 'utf-8');
      const data: ModuleRegistryRaw = JSON.parse(raw);

      if (!data || !data.domains) {
        console.warn('[ModuleRegistry] 注册表格式错误');
        return false;
      }

      const modules = new Map<string, ModuleInfo>();
      const domains = data.domains;

      for (const [domainKey, domainData] of Object.entries(domains)) {
        const categories = domainData.categories || {};
        for (const [catKey, catData] of Object.entries(categories)) {
          const modMap = catData.modules || {};
          for (const [modKey, modData] of Object.entries(modMap)) {
            const info: ModuleInfo = {
              ...modData,
              domain: domainKey,
              category_name: catKey,
            } as ModuleInfo;
            modules.set(info.id, info);
          }
        }
      }

      this.rawData = data;
      this.modules = modules;
      this.lastLoadTime = Date.now();
      this.lastFileMtime = fileMtime;
      this.buildIndexes();

      console.log(`[ModuleRegistry] 加载成功，共 ${modules.size} 个模块`);
      return true;
    } catch (e) {
      console.error('[ModuleRegistry] 加载失败:', e);
      return false;
    }
  }

  private buildIndexes(): void {
    this.byChain.clear();
    this.byCategory.clear();
    this.byDomain.clear();
    this.byStage.clear();
    this.byTag.clear();

    for (const [id, mod] of this.modules) {
      if (!this.byChain.has(mod.chain)) {
        this.byChain.set(mod.chain, new Set());
      }
      this.byChain.get(mod.chain)!.add(id);

      if (!this.byCategory.has(mod.category)) {
        this.byCategory.set(mod.category, new Set());
      }
      this.byCategory.get(mod.category)!.add(id);

      if (!this.byDomain.has(mod.domain)) {
        this.byDomain.set(mod.domain, new Set());
      }
      this.byDomain.get(mod.domain)!.add(id);

      for (const stage of mod.applicable_stages) {
        if (!this.byStage.has(stage)) {
          this.byStage.set(stage, new Set());
        }
        this.byStage.get(stage)!.add(id);
      }

      for (const tag of mod.tags) {
        if (!this.byTag.has(tag)) {
          this.byTag.set(tag, new Set());
        }
        this.byTag.get(tag)!.add(id);
      }
    }
  }

  reload(): boolean {
    return this.load();
  }

  checkForUpdates(): boolean {
    try {
      if (!fs.existsSync(this.registryPath)) {
        return false;
      }
      const stat = fs.statSync(this.registryPath);
      if (stat.mtimeMs > this.lastFileMtime) {
        console.log('[ModuleRegistry] 检测到文件更新，重新加载...');
        return this.load();
      }
      return false;
    } catch {
      return false;
    }
  }

  startWatching(intervalMs: number = 5000): void {
    if (this.watchInterval) return;

    this.watchInterval = setInterval(() => {
      this.checkForUpdates();
    }, intervalMs);

    console.log(`[ModuleRegistry] 文件监听已启动，间隔 ${intervalMs}ms`);
  }

  stopWatching(): void {
    if (this.watchInterval) {
      clearInterval(this.watchInterval);
      this.watchInterval = null;
    }
  }

  // ============================================================
  // 查询接口
  // ============================================================

  get(moduleId: string): ModuleInfo | undefined {
    return this.modules.get(moduleId);
  }

  has(moduleId: string): boolean {
    return this.modules.has(moduleId);
  }

  getAll(): ModuleInfo[] {
    return Array.from(this.modules.values());
  }

  count(): number {
    return this.modules.size;
  }

  query(params: ModuleQueryParams): ModuleInfo[] {
    let candidates = new Set(this.modules.keys());

    if (params.chain) {
      const chains = Array.isArray(params.chain) ? params.chain : [params.chain];
      const chainSet = new Set<string>();
      for (const c of chains) {
        const s = this.byChain.get(c);
        if (s) s.forEach(id => chainSet.add(id));
      }
      if (chainSet.size > 0) {
        candidates = new Set([...candidates].filter(x => chainSet.has(x)));
      }
    }

    if (params.category) {
      const cats = Array.isArray(params.category) ? params.category : [params.category];
      const catSet = new Set<string>();
      for (const c of cats) {
        const s = this.byCategory.get(c);
        if (s) s.forEach(id => catSet.add(id));
      }
      if (catSet.size > 0) {
        candidates = new Set([...candidates].filter(x => catSet.has(x)));
      }
    }

    if (params.domain) {
      const domains = Array.isArray(params.domain) ? params.domain : [params.domain];
      const domainSet = new Set<string>();
      for (const d of domains) {
        const s = this.byDomain.get(d);
        if (s) s.forEach(id => domainSet.add(id));
      }
      if (domainSet.size > 0) {
        candidates = new Set([...candidates].filter(x => domainSet.has(x)));
      }
    }

    if (params.stage) {
      const stages = Array.isArray(params.stage) ? params.stage : [params.stage];
      const stageSet = new Set<string>();
      for (const st of stages) {
        const s = this.byStage.get(st);
        if (s) s.forEach(id => stageSet.add(id));
      }
      if (stageSet.size > 0) {
        candidates = new Set([...candidates].filter(x => stageSet.has(x)));
      }
    }

    if (params.tag) {
      const tags = Array.isArray(params.tag) ? params.tag : [params.tag];
      const tagSet = new Set<string>();
      for (const t of tags) {
        const s = this.byTag.get(t);
        if (s) s.forEach(id => tagSet.add(id));
      }
      if (tagSet.size > 0) {
        candidates = new Set([...candidates].filter(x => tagSet.has(x)));
      }
    }

    const results: ModuleInfo[] = [];
    for (const id of candidates) {
      const mod = this.modules.get(id)!;

      if (params.security_level && mod.security_level !== params.security_level) {
        continue;
      }

      if (params.min_accuracy !== undefined && mod.historical_accuracy < params.min_accuracy) {
        continue;
      }

      if (params.max_tokens !== undefined && mod.estimated_tokens > params.max_tokens) {
        continue;
      }

      if (params.intent && !mod.applicable_intents.includes(params.intent)) {
        continue;
      }

      if (params.market_condition && !mod.market_conditions.includes(params.market_condition)) {
        continue;
      }

      results.push(mod);
    }

    return results;
  }

  getByChain(chain: ModuleInfo['chain']): ModuleInfo[] {
    return this.query({ chain });
  }

  getByDomain(domain: string): ModuleInfo[] {
    return this.query({ domain });
  }

  getDependencies(moduleId: string): ModuleInfo[] {
    const mod = this.get(moduleId);
    if (!mod) return [];
    return mod.dependencies
      .map(id => this.get(id))
      .filter((m): m is ModuleInfo => m !== undefined);
  }

  getFallback(moduleId: string): ModuleInfo | undefined {
    const mod = this.get(moduleId);
    if (!mod || !mod.fallback?.enabled) return undefined;
    const fallbackId = mod.fallback.fallback_module;
    if (!fallbackId) return undefined;
    return this.get(fallbackId);
  }

  getRaw(): ModuleRegistryRaw | null {
    return this.rawData;
  }

  getDomains(): string[] {
    return Array.from(this.byDomain.keys());
  }

  getChains(): string[] {
    return Array.from(this.byChain.keys());
  }

  getCategories(): string[] {
    return Array.from(this.byCategory.keys());
  }

  getStats() {
    return {
      total: this.modules.size,
      byChain: Object.fromEntries(
        Array.from(this.byChain.entries()).map(([k, v]) => [k, v.size])
      ),
      byDomain: Object.fromEntries(
        Array.from(this.byDomain.entries()).map(([k, v]) => [k, v.size])
      ),
      byCategory: Object.fromEntries(
        Array.from(this.byCategory.entries()).map(([k, v]) => [k, v.size])
      ),
      lastLoad: this.lastLoadTime,
    };
  }

  isActive(moduleId: string): boolean {
    const mod = this.get(moduleId);
    if (!mod) return false;
    return mod.lifecycle?.status === 'active' && !mod.lifecycle?.deprecated;
  }

  getRegistryPath(): string {
    return this.registryPath;
  }
}

// ============================================================
// 单例
// ============================================================

let globalRegistry: ModuleRegistry | null = null;

export function getModuleRegistry(): ModuleRegistry {
  if (!globalRegistry) {
    globalRegistry = new ModuleRegistry();
  }
  return globalRegistry;
}

export function reloadModuleRegistry(): boolean {
  return getModuleRegistry().reload();
}
